Added a flake for building Prusti with the Nix package manager

jmc-figueira opened the pull request:
This flake allows one to build Prusti through the Nix package manager
(https://nixos.org/), on NixOS or other distros with Nix installed,
as well as macOS.

Caveats:
- Currently the flake will be built from source every time, which takes
  a considerable amount of time. A cache setup through https://app.cachix.org/
  may be used to solve this problem (and can be setup as a CI workflow).
- As of 04/01/2021, rustfmt was not available in the nightly (as per
  https://rust-lang.github.io/rustup-components-history/). The
  `rust-toolchain` file was altered to default to a minimal profile,
  with optional components added explicitly. If this is unacceptable, an
  alternative must be found, as toolchain component unavailability in
  Nix is an error (aborts the build) and not a warning.
- Currently, Viper is simply downloaded from the official Nightly
  download link, as the Python script does. This causes a problem, as
  Nix builds must be hermetic and reproducible, and changes in the Viper
  nightly, which is unversioned, cause the SHA256 to change and abort
  the build. A better solution would be to compile Viper from source
  using a separate Nix expression or to download a version-pinned
  archive (if possible).
- Currently, Prusti is not agnostic to the Rust toolchain instalation
  method and expects `rustup` to be used. This was an issue on runtime
  for `prusti-launch`, which expects to find the Rust sysroot through
  `rustup`. This was altered by allowing an environment variable,
  `PRUSTI_SYSROOT`, to be set, pointing to the Rust sysroot in an
  environment-agnostic manner, which makes it compatible with the way
  the Rust toolchain is installed as a dependency in Nix (through
  https://github.com/oxalica/rust-overlay). The developer documentation
  regarding environment variables should probably be updated to account
  for this
  (https://viperproject.github.io/prusti-dev/dev-guide/development/build.html).
- Documentation should, similarly, be updated to provide instructions
  on installing and setting up a development environment through Nix.

Commit:
Added a flake for building Prusti with the Nix package manager

This flake allows one to build Prusti through the Nix package manager
(on NixOS or other distros with Nix installed, as well as macOS).

Caveats:
- Currently the flake will be built from source every time, which takes
  a considerable amount of time. A cache setup through https://app.cachix.org/
  may be used to solve this problem (and can be setup as a CI workflow).
- As of 04/01/2021, rustfmt was not available in the nightly (as per
  https://rust-lang.github.io/rustup-components-history/). The
  `rust-toolchain` file was altered to default to a minimal profile,
  with optional components added explicitly. If this is unacceptable, an
  alternative must be found, as toolchain component unavailability in
  Nix is an error (aborts the build) and not a warning.
- Currently, Viper is simply downloaded from the official Nightly
  download link, as the Python script does. This causes a problem, as
  Nix builds must be hermetic and reproducible, and changes in the Viper
  nightly, which is unversioned, cause the SHA256 to change and abort
  the build. A better solution would be to compile Viper from source
  using a separate Nix expression or to download a version-pinned
  archive (if possible).
- Currently, Prusti is not agnostic to the Rust toolchain instalation
  method and expects `rustup` to be used. This was an issue on runtime
  for `prusti-launch`, which expects to find the Rust sysroot through
  `rustup`. This was altered by allowing an environment variable,
  `PRUSTI_SYSROOT`, to be set, pointing to the Rust sysroot in an
  environment-agnostic manner, which makes it compatible with the way
  the Rust toolchain is installed as a dependency in Nix (through
  https://github.com/oxalica/rust-overlay). The developer documentation
  regarding environment variables should probably be updated to account
  for this
  (https://viperproject.github.io/prusti-dev/dev-guide/development/build.html).
- Documentation should, similarly, be updated to provide instructions
  on installing and setting up a development environment through Nix.

rustfmt temporarily removed due to build failure

Tool availability may be checked here: https://rust-lang.github.io/rustup-components-history/

Added a wrapper to set the necessary runtime variables for the flake

Patched Prusti to accept non-Rustup toolchains at runtime

Useful for Nix Rust installations, for example

Moved sysroot obtaining logic to the appropriate function

As requested by https://github.com/viperproject/prusti-dev/pull/464\#discussion_r615806829

Added libs to the derivation output

Needed for `prusti_contracts` to work

fpoli commented on prusti-launch/src/bin/prusti-rustc.rs:
```diff
@@ -36,8 +36,11 @@ fn process(mut args: Vec<String>) -> Result<(), i32> {
+        Ok(prusti_sysroot) => PathBuf::from(prusti_sysroot),
+        Err(_) => prusti_launch::prusti_sysroot()
+            .expect(&format!("Failed to find Rust's sysroot for Prusti"))
+    };
```
Could you move this match into `prusti_sysroot`?

jmc-figueira replied:
Done in https://github.com/viperproject/prusti-dev/pull/464/files#diff-dd855b3156f3186827e7c4436f3c131a7d57f9cbba9616567a0522412c8d3eb5, please check if it is ok.

fpoli commented on flake.nix:
```diff
@@ -0,0 +1,90 @@
+          name = "viper";
+          url = "https://viper.ethz.ch/downloads/ViperToolsNightlyLinux.zip";
+          stripRoot = false;
+          hash = "sha256-oV9VgGmzNrLW6n08vXU2rlV6S9Eca3U0L+gNbVxKhEE=";
```
`https://viper.ethz.ch/downloads/ViperToolsNightlyLinux.zip` is overwritten often. `http://viper.ethz.ch/downloads/ViperToolsLastReleaseLinux.zip` might a bit more stable, but I haven't tried it. We are still working on a way to refer to an exact old nightly version of Viper.

jmc-figueira replied:
A possible solution suggested by @vakaras in Zulip (https://prusti.zulipchat.com/#narrow/stream/236614-general/topic/Building.20Prusti.20on.20NixOS/near/229701863), at least for Nix, would be to build Viper from source using a separate Nix flake in its repository, which I started working on as well (will reference the relevant pull request when it's ready). It would then be possible to pin the dependency to a specific commit (done automatically by `flake.lock`, in fact), which can be updated as necessary by running `nix flake update`. The only caveat with this approach, as of now, would be the added time required to compile Viper in addition to Prusti. This should be resolved once a cache is hooked up, where it would be possible to have a workflow automatically compiling binaries of both Prusti and Viper on each commit to master, which will be made available to users by the Nix caching system.

Commit:
Changed environment variable to be consistent with tests and sanitized flake

The new `PRUSTI_SYSROOT` environment variable was changed to match one previously used by tests, `RUST_SYSROOT`

The structure of the flake was also sanitized, for better readability

Commit:
Fixed flake so the necessary libraries are copied to the package

Followed the Dockerfile's approach to determine which libraries are required

Commit:
Added a simple test

Should prevent further issues with erroneous build results. This was based on the Dockerfile's test; running `cargo test`, as well as more complex tests (possibly relying on `test-crates`) should also be added

vakaras reviewed (approved):
LGTM Thanks!

@fpoli Unless you object, I will merge this in.
